src/main/scala/net/snowflake/spark/snowflake/ServerConnection.scala

```diff
@@ -0,0 +1,290 @@
+      case "" => true
+      case createTablePattern(_, _, _, _, _) => true
+      case createStagePattern(_, _, _, _) => true
+      case dropObjects(_, _) => true
+      case mergeInto(_) => true
+      case _ => false
+    }
+  }
+
+  def isConnectionCacheSupported: Boolean = {
+    // Support sharing connection if pre/post actions doesn't change context
+    ServerConnection.supportSharingJDBCConnection &&
+      parameters.supportShareConnection &&
+      parameters.preActions.forall(isQueryInWhiteList) &&
+      parameters.postActions.forall(isQueryInWhiteList)
+  }
+}
+
+private[snowflake] object ServerConnection {
+  private val log = LoggerFactory.getLogger(getClass)
+  private[snowflake] val serverConnectionCount = new AtomicLong
+  private[snowflake] val jdbcConnectionCount = new AtomicLong
+  private[snowflake] var supportSharingJDBCConnection = true
+
+  private[snowflake] def setSupportSharingJDBCConnection(enabled: Boolean): Unit =
+    supportSharingJDBCConnection = enabled
+
+  def apply(jdbcConnection: Connection, enableCache: Boolean): ServerConnection =
+    new ServerConnection(jdbcConnection, enableCache)
+
+  def apply(jdbcConnection: Connection): ServerConnection =
+    new ServerConnection(jdbcConnection, true)
+
+  private val cachedJdbcConnections = mutable.Map[ConnectionCacheKey, Connection]()
+
+  def closeAllCachedConnections: Unit = synchronized {
+    log.info(s"Close all ${cachedJdbcConnections.size} cached connection.")
+    cachedJdbcConnections.values.foreach(_.close())
+    cachedJdbcConnections.clear()
+  }
+
+  def getServerConnection(parameters: MergedParameters, useCache: Boolean = true)
+  : ServerConnection = synchronized {
+    val connectionCacheKey = new ConnectionCacheKey(parameters)
+    val enableCache = useCache && connectionCacheKey.isConnectionCacheSupported
+    val (jdbcConnection, newConnection) =
+      if (enableCache && cachedJdbcConnections.keySet.contains(connectionCacheKey)) {
+        (cachedJdbcConnections(connectionCacheKey), false)
+      } else {
+        val createdJdbcConnection = createJDBCConnection(parameters)
+        if (enableCache) {
+          log.debug(s"Cache the new created JDBCConnection")
+          cachedJdbcConnections.put(connectionCacheKey, createdJdbcConnection)
+        }
+        (createdJdbcConnection, true)
+      }
+    val serverConnection = ServerConnection(jdbcConnection, enableCache)
+    log.info(s"Create ServerConnection with ${if (newConnection) "new" else "cached"}" +
+      s" JDBC connection: ${serverConnection.getSessionID}")
+
+    // Send client info telemetry message.
+    if (newConnection) {
+      val extraValues = Map(TelemetryClientInfoFields.SFURL -> parameters.sfURL,
+        TelemetryClientInfoFields.SHARED -> enableCache.toString)
+      SnowflakeTelemetry.sendClientInfoTelemetry(extraValues, serverConnection)
+    }
+
+    serverConnection
+  }
+
+  private def createJDBCConnection(params: MergedParameters): Connection = {
```

Comment:
This function is moved from `DefaultJDBCWrapper.getConnector()`. The only extra change is to move below to `getServerConnection()`
```
    // Send client info telemetry message.
    if (newConnection) {
      val extraValues = Map(TelemetryClientInfoFields.SFURL -> parameters.sfURL,
        TelemetryClientInfoFields.SHARED -> enableCache.toString)
      SnowflakeTelemetry.sendClientInfoTelemetry(extraValues, serverConnection)
    }
```